keepassxc-browser/background/client.js

```diff
@@ -0,0 +1,333 @@
+                this.buffer.splice(i, 1);
+                return true;
+            }
+        }
+
+        return false;
+    }
+};
+
+//--------------------------------------------------------------------------
+// Messaging
+//--------------------------------------------------------------------------
+
+keepassClient.sendNativeMessage = function(request, enableTimeout = false, timeoutValue) {
+    return new Promise((resolve, reject) => {
+        let timeout;
+        const requestAction = request.action;
+        const ev = keepassClient.nativePort.onMessage;
+
+        const listener = ((port, action) => {
+            const handler = (msg) => {
+                if (msg && msg.action === action) {
+                    // Only resolve a matching response or a notification (without nonce)
+                    if (!msg.nonce || messageBuffer.matchAndRemove(msg)) {
+                        port.removeListener(handler);
+                        if (enableTimeout) {
+                            clearTimeout(timeout);
+                        }
+
+                        resolve(msg);
+                        return;
+                    }
+                }
+            };
+            return handler;
+        })(ev, requestAction);
+        ev.addListener(listener);
+
+        const messageTimeout = timeoutValue || keepassClient.messageTimeout;
+
+        // Handle timeouts
+        if (enableTimeout) {
+            timeout = setTimeout(() => {
+                const errorMessage = {
+                    action: requestAction,
+                    error: kpErrors.getError(kpErrors.TIMEOUT_OR_NOT_CONNECTED),
+                    errorCode: kpErrors.TIMEOUT_OR_NOT_CONNECTED
+                };
+                keepass.isKeePassXCAvailable = false;
+                ev.removeListener(listener.handler);
+                resolve(errorMessage);
+            }, messageTimeout);
+        }
+
+        // Store the request to the buffer
+        messageBuffer.addMessage(request);
+
+        // Send the request
+        if (keepassClient.nativePort) {
+            keepassClient.nativePort.postMessage(request);
+        }
+    });
+};
+
+keepassClient.handleResponse = function(response, incrementedNonce, tab) {
+    if (response.message && response.nonce) {
+        const res = keepassClient.decrypt(response.message, response.nonce);
+        if (!res) {
+            keepass.handleError(tab, kpErrors.CANNOT_DECRYPT_MESSAGE);
+            return [ false, undefined ];
+        }
+
+        const message = nacl.util.encodeUTF8(res);
+        const parsed = JSON.parse(message);
+
+        if (keepassClient.verifyResponse(parsed, incrementedNonce)) {
+            return [ true, parsed ];
+        }
+    } else if (response.error && response.errorCode) {
+        keepass.handleError(tab, response.errorCode, response.error);
+        return [ false, undefined ];
+    }
+
+    return [ false, undefined ];
```

Comment:
Should just return `undefined` then test for that to see if the response is valid.